### PR TITLE
Feature/lc -2 memory without langchain

### PR DESCRIPTION
**Pull request by superjkm**
This implements context memory for the LLM by accumulating the messages sent to and received from the LLM.

There is a verbose mode, passing 'true' into the instantiation of the memory object. 

*Note - this code is server side currently, but could be moved to client side if that's a better way to segregate - please feed me back!

### CODE REVIEW

**Review by jcm5155 — APPROVED** [review]

**Review by scaperoth — APPROVED** [review]
Some 🥫 for 🧠 

Looks good!

File: frontend/labs-chat/src/helpers/memory.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+     addHumanMessage (message) {
+        // todo: add a limit to the context size, plus auto reduce context size
+        if (this.verbose) {
+            console.log(`Adding human message: ${message}`);
+        }
+        this.history += this.human_prefix + message + "\n\n";
+    }
+
+    addAIMessage = function () {
```
[the remark this answers is not in the document]
Comment (NAPTlME) [reply frontend/labs-chat/src/helpers/memory.js]: @scaperoth Does commitAIStream work? (to follow along with `accumulateAIStream` and `clearAIStream`)

File: frontend/labs-chat/src/helpers/memory.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+     addHumanMessage (message) {
+        // todo: add a limit to the context size, plus auto reduce context size
+        if (this.verbose) {
+            console.log(`Adding human message: ${message}`);
+        }
+        this.history.push({role: this.human_role, content: message});
+    }
+
+    commitAIStream = function () {
```
Comment (superjkm) [inline frontend/labs-chat/src/helpers/memory.js:19]:
```suggestion
    commitAIStream() {
```

File: frontend/labs-chat/src/helpers/memory.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        // todo: add a limit to the context size, plus auto reduce context size
+        if (this.verbose) {
+            console.log(`Adding AI message: ${this.ai_stream}`);
+        }
+        this.history.push({role: this.ai_role, content: this.ai_stream});
+        // clear the stream upon committing
+        this.clearAIStream();
+    };
+
+    accumulateAIStream = function (textDelta) {
```
Comment (superjkm) [inline frontend/labs-chat/src/helpers/memory.js:29]:
```suggestion
    accumulateAIStream(textDelta) {
```

File: frontend/labs-chat/src/helpers/memory.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        if (this.verbose) {
+            console.log(`Accumulating AI stream: ${textDelta}`);
+        }
+        this.ai_stream += textDelta;
+    }
+
+    clearAIStream = function () {
```
Comment (superjkm) [inline frontend/labs-chat/src/helpers/memory.js:36]:
```suggestion
    clearAIStream() {
```

File: frontend/labs-chat/src/helpers/memory.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+export class Memory {
+    constructor(verbose = false) {
+        this.history = [];
+        this.ai_stream = "";
+        this.verbose = verbose;
+        this.human_role = "user";
+        this.ai_role = "assistant";
+    }
+
+     addHumanMessage (message) {
```
Comment (superjkm) [inline frontend/labs-chat/src/helpers/memory.js:11]:
```suggestion
     addHumanMessage(message) {
```

File: frontend/labs-chat/src/pages/api/streaming-llm.js [hunk 6]
```diff
@@ -46,12 +62,14 @@ export default async function handler(req, res) {
 
   try {
     const request = await req.json();
-    const input = {role: "user", content: request.input};
+    //append the human message to the context
+    memory.addHumanMessage(request.input)
+    console.log(`history: ${JSON.stringify(memory.history)}`);
```
Comment (superjkm) [inline frontend/labs-chat/src/pages/api/streaming-llm.js:67]:
```suggestion
```

**Review by NAPTlME — APPROVED** [review]

**Pull request merged**